refactor comparison support

Lojjs opened the pull request.

Lojjs committed:
Refactored `updateWith` in `CypherComparisonSupport`

Lojjs committed:
Added negative testing in CypherComparisonSupport

Lojjs committed:
Fix nullpointer from test

RewindableExecutionResult sets prefix as null from older versions

sherfert commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/CypherComparisonSupport.scala:
```diff
@@ -95,209 +95,153 @@ trait CypherComparisonSupport extends CypherTestSupport {
-    if (hasOverlap)
-      fail("Set of configurations that were expected to succeed and the set of configs that should be ignored are overlapping")
+  private def extractFirstScenario(config: TestConfiguration, compareResults: TestConfiguration, comparePlans: TestConfiguration): TestScenario = {
+    val scenariosToChooseFrom = (compareResults.scenarios.isEmpty, comparePlans.scenarios.isEmpty) match {
```
Since we are assuming `compareResults` and `comparePlans` are a subset of `config`,
can we not always intersect all three (and fail if that intersection is empty)?

OliviaYtterbrink replied:
Sometimes the plans aren't comparable but the results are, intersecting these we would then end up with an empty set when we just want the intersection between `compareResults` and `config`

sherfert commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/CypherComparisonSupport.scala:
```diff
@@ -95,209 +95,153 @@ trait CypherComparisonSupport extends CypherTestSupport {
-    val hasOverlap = expectedSuccessFrom.scenarios.exists(ignoreScenarios.scenarios.contains(_))
-    if (hasOverlap)
-      fail("Set of configurations that were expected to succeed and the set of configs that should be ignored are overlapping")
+  private def extractFirstScenario(config: TestConfiguration, compareResults: TestConfiguration, comparePlans: TestConfiguration): TestScenario = {
```
Can we change the parameter name `config` to `expectedSuccessFrom` or similar?

sherfert commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/CypherComparisonSupport.scala:
```diff
@@ -95,209 +95,153 @@ trait CypherComparisonSupport extends CypherTestSupport {
     firstScenario.prepare()
     executeBefore()
-    val lastResult = innerExecute(s"CYPHER ${firstScenario.preparserOptions} $query", params.toMap)
+    val lastResult = innerExecute(s"CYPHER ${firstScenario.preparserOptions} $query", params)
```
`firstScenario` and `lastResult` seem inconsistent. They're both the scenario/result you're comparing against, aren't they?

OliviaYtterbrink replied:
👍

sherfert commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/CypherComparisonSupport.scala:
```diff
@@ -95,209 +95,153 @@ trait CypherComparisonSupport extends CypherTestSupport {
+    val firstScenario = extractFirstScenario(expectedSuccessFrom, compareResults, comparePlans)
+
+    // let this throw to enable catching exceptions outside
+    firstScenario.prepare()
```
I commented this block out and ran all acceptance and all compatibility tests. No failures. Can you point me to the test(s) that require this to be here?

OliviaYtterbrink replied:
`test("should refuse to execute index seeks using inequalities over different types")` in `NodeIndexSeekByRangeAcceptanceTest` is reliant on this being here (when it has been reverted to actually check for the exception again)

sherfert commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/AggregationAcceptanceTest.scala:
```diff
@@ -40,8 +42,9 @@ class AggregationAcceptanceTest extends ExecutionEngineFunSuite with CypherCompa
                    |RETURN id(selectedFriendship) AS friendshipId, selectedFriendship.propFive AS propertyValue""".stripMargin
     val params = "param" -> 3
```
Now that `executeWith` accepts a map of params, we should probably define this right here as a map already.

sherfert commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/CypherComparisonSupport.scala:
```diff
@@ -457,6 +405,8 @@ object CypherComparisonSupport {
 
   object Runtimes {
 
+    val all = Runtimes(CompiledBytecode, CompiledSource, Slotted, Interpreted, ProcedureOrSchema, Default)
```
In `Planner.all` and `Versions.all` Alex and I explicitly excluded `Default`, since one of them would be the default anyway. Is that required to be included here?

OliviaYtterbrink replied:
There is no `Versions.all`, and `Default` in runtimes is actually different from specifying one

sherfert commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/NodeIndexSeekByRangeAcceptanceTest.scala:
```diff
@@ -928,19 +926,14 @@ class NodeIndexSeekByRangeAcceptanceTest extends ExecutionEngineFunSuite with Cy
 
     val query = "MATCH (n:Label) WHERE n.prop >= '1' AND n.prop > 10 RETURN n.prop AS prop"
 
-    an[IllegalArgumentException] should be thrownBy {
```
Is that `IllegalArgumentException` not thrown anymore or do we not assert it is thrown any more?

OliviaYtterbrink replied:
this was wrongly removed

sherfert reviewed (changes requested):
This is a nice step forward, I like it! The comments below are just minor things that can enhance readability of the code or questions.

OliviaYtterbrink committed:
Cleanup after comments

Better named variables
Helper to check if a TestConfiguration contains a scenario
Allow specifying that no scenarios should succeed

sherfert committed:
Added a comment for clarity
